File: baekjoon/p1005.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

struct Node {
    int b_time = 0;
    int degree = 0;
    vector<int> next;

    bool operator>(Node n) const {
        return b_time > n.b_time;
    }
};

void work() {
    int b_num, r_num;
    cin >> b_num >> r_num;

    vector<Node> nodes;

    for (int i = 0; i < b_num; i++) {
        int b_time;
        cin >> b_time;
        Node node;
        node.b_time = b_time;
        nodes.push_back(node);
    }

    nodes.resize(b_num);

    for (int i = 0; i < r_num; i++) {
        int start, end;
        cin >> start >> end;

        start--;
        end--;

        nodes[start].next.push_back(end);
        nodes[end].degree++;
    }

    int final_b;
    cin >> final_b;
    final_b--;

    priority_queue<Node, vector<Node>, greater<Node>> queue;
    int finish_time = 0;

    for (int i = 0; i < b_num; i++) {
        if (nodes[i].degree == 0) {
            queue.push(nodes[i]);
        }
    }

    while (!queue.empty()) {
        Node cur = queue.top();
        queue.pop();

        bool finished = false;
        vector<int>::iterator iter = cur.next.begin();
        for (; iter != cur.next.end(); iter++) {
            Node *next_node = &nodes[*iter];
            next_node->degree--;
            if (next_node->degree == 0) {
                next_node->b_time += cur.b_time;
                if (*iter == final_b) {
                    finish_time = next_node->b_time;
                    finished = true;
                    break;
                }
                queue.push(*next_node);
            }
        }
        if (finished) {
            break;
        }
    }

    cout << finish_time << endl;
}

int main() {
    ios_base::sync_with_stdio(false);
    int test_case;

    cin >> test_case;
    for (int i = 0; i < test_case; i++) {
        work();
    }
}

File: baekjoon/p1005.py
import sys
import heapq


class Node(object):
    def __init__(self):
        self.degree = 0
        self.next = list()
        self.b_time = 0

def work():
    b_num, r_num = map(int, sys.stdin.readline().rstrip().split())
    
    rules = [Node() for _ in range(b_num)]

    for i, t in enumerate(map(int, sys.stdin.readline().rstrip().split())):
        rules[i].b_time = t

    for _ in range(r_num):
        start, end = map(int, sys.stdin.readline().rstrip().split())

        # index is started from 0
        start -= 1
        end -= 1

        rules[start].next.append(end)
        rules[end].degree += 1

    # index is started from 0
    final_b = int(sys.stdin.readline())
    final_b -= 1

    heap = list()
    for n in rules:
        if n.degree == 0:
            heapq.heappush(heap, (n.b_time, n))

    final_b_time = 0

    while heap:
        node_time, node = heapq.heappop(heap)
        finished = False
        for n_idx in node.next:
            next_node = rules[n_idx]
            next_node.degree -= 1
            if next_node.degree == 0:
                next_node.b_time += node_time
                if n_idx == final_b:
                    final_b_time = next_node.b_time
                    finished = True
                    break
                heapq.heappush(heap, (next_node.b_time, next_node))
        if finished:
            break

    sys.stdout.write('%d\n' % final_b_time)


test_case = int(sys.stdin.readline())

for _ in range(test_case):
    work()
